File: Java/0417_Atlantic_Pacific_Water_Flow.java
class Solution {
    public int[][] dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    public int m, n;
    public List<int[]> pacificAtlantic(int[][] matrix) {
        List<int[]> ans = new ArrayList<int[]>();
        m = matrix.length;
        if (m == 0) {
            return ans;
        }
        n = matrix[0].length;
        boolean[][] pacifica = new boolean[m][n], atlantic = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            if (!pacifica[i][0]) {
                pacifica[i][0] = true;
                dfs(pacifica, matrix, i, 0);
            }
            if (!atlantic[i][n - 1]) {
                atlantic[i][n - 1] = true;
                dfs(atlantic, matrix, i, n - 1);
            }
        }
        for (int j = 0; j < n; j++) {
            if (!pacifica[0][j]) {
                pacifica[0][j] = true;
                dfs(pacifica, matrix, 0, j);
            }
            if (!atlantic[m - 1][j]) {
                atlantic[m - 1][j] = true;
                dfs(atlantic, matrix, m - 1, j);
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (atlantic[i][j] && pacifica[i][j]) {
                    int[] coord = {i, j};
                    ans.add(coord);
                }
            }
        }
        return ans;
    }
    public void dfs(boolean[][] dp, int[][] matrix, int i, int j) {
        for (int[] dir: dirs) {
            int ni = i + dir[0], nj = j + dir[1];
            if (-1 < ni && ni < m && -1 < nj && nj < n && !dp[ni][nj] && matrix[i][j] <= matrix[ni][nj]) {
                dp[ni][nj] = true;
                dfs(dp, matrix, ni, nj);
            }
        }
    }
}

File: Python3/0417_Atlantic_Pacific_Water_Flow.py
class Solution:
    def pacificAtlantic(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[List[int]]
        """
        def pr(dp):
            for row in dp:
                print(row)
        if not matrix:
            return []
        ans = []
        dirs = [(0, 1), (1, 0), (-1, 0), (0, -1)]
        m, n = len(matrix), len(matrix[0])
        pacifica = [[False for j in range(n)] for i in range(m)]
        atlantic = [[False for j in range(n)] for i in range(m)]
        def dfs(dp, i, j):            
            for ni, nj in dirs:
                ni, nj = i + ni, j + nj
                if -1 < ni < m and -1 < nj < n and not dp[ni][nj] and matrix[i][j] <= matrix[ni][nj]:
                    dp[ni][nj] = True
                    dfs(dp, ni, nj)
        for j in range(n):
            if not pacifica[0][j]:
                pacifica[0][j] = True
                dfs(pacifica, 0, j)
            if not atlantic[-1][j]:
                atlantic[-1][j] = True
                dfs(atlantic, m - 1, j)
        for i in range(m):
            if not pacifica[i][0]:
                pacifica[i][0] = True
                dfs(pacifica, i, 0)
            if not atlantic[i][-1]:
                atlantic[i][-1] = True
                dfs(atlantic, i, n - 1)
        for i in range(m):
            for j in range(n):
                if atlantic[i][j] and pacifica[i][j]:
                    ans.append([i, j])
        return ans
